### ISSUE TITLE
Armory doesn't launch on Mac 10.12.2

### ISSUE THREAD

**Issue by utdrmac** [goatpig/BitcoinArmory#163]
I just downloaded 0.95.1_osx.tar.gz and attempted to launch Armory.app. Nothing happens. I can see the dock icon for a split second then it goes away. No error messages on screen.

I found this in Console.app:
 Jan 12 11:47:40 othala com.apple.xpc.launchd[1] (com.armory.armory.13444[18903]): Service exited with abnormal code: 126

**Comment by utdrmac** [comment]
Tested with 0.93.3. Same issue/error code.

**Comment by utdrmac** [comment]
Happy to help debug.

**Comment by utdrmac** [comment]
#161 Possible issue?

**Comment by TheCranston** [comment]
I can confirm it does not launch.  A manual launch of armoryd yields:

`
$ ./armoryd 
ERROR:root:code for hash md5 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type md5
ERROR:root:code for hash sha1 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type sha1
ERROR:root:code for hash sha224 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type sha224
ERROR:root:code for hash sha256 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type sha256
ERROR:root:code for hash sha384 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type sha384
ERROR:root:code for hash sha512 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type sha512
Traceback (most recent call last):
  File "/Users/Me/Downloads/Armory 2.app/Contents/MacOS/py/usr/lib/armory/armoryd.py", line 100, in <module>
    from twisted.cred.checkers import FilePasswordDB
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/cred/checkers.py", line 14, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/cred/credentials.py", line 22, in <module>
ImportError: cannot import name md5
`

And a manual launch of Armory results in:

`
$ ./Armory 
ERROR:root:code for hash md5 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type md5
ERROR:root:code for hash sha1 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type sha1
ERROR:root:code for hash sha224 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type sha224
ERROR:root:code for hash sha256 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type sha256
ERROR:root:code for hash sha384 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type sha384
ERROR:root:code for hash sha512 was not found.
Traceback (most recent call last):
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
ValueError: unsupported hash type sha512
INFO:root:ArmoryUtils.py:1137 - C++ block utilities loaded successfully
INFO:root:ArmoryUtils.py:651 - Executing popen: sysctl hw.memsize
INFO:root:ArmoryUtils.py:651 - Executing popen: sysctl -n machdep.cpu.brand_string
INFO:root:ArmoryUtils.py:1247 - 
INFO:root:ArmoryUtils.py:1248 - 
INFO:root:ArmoryUtils.py:1249 - 
INFO:root:ArmoryUtils.py:1250 - ************************************************************
INFO:root:ArmoryUtils.py:1251 - Invoked: /Users/Me/Downloads/Armory 2.app/Contents/MacOS/py/usr/lib/armory/ArmoryQt.py
INFO:root:ArmoryUtils.py:1252 - ************************************************************
INFO:root:ArmoryUtils.py:1253 - Loading Armory Engine:
INFO:root:ArmoryUtils.py:1254 -    Armory Version        : 0.95.1
INFO:root:ArmoryUtils.py:1255 -    Armory Build:         : None
INFO:root:ArmoryUtils.py:1256 -    PyBtcWallet  Version  : 1.35
INFO:root:ArmoryUtils.py:1257 - Detected Operating system: MacOSX
INFO:root:ArmoryUtils.py:1258 -    OS Variant            : 10.12.2
INFO:root:ArmoryUtils.py:1259 -    User home-directory   : /Users/Me/Library/Application Support
INFO:root:ArmoryUtils.py:1260 -    Satoshi BTC directory : /Users/Me/Library/Application Support/Bitcoin/
INFO:root:ArmoryUtils.py:1261 -    Armory home dir       : /Users/Me/Library/Application Support/Armory/
INFO:root:ArmoryUtils.py:1262 - Detected System Specs    : 
INFO:root:ArmoryUtils.py:1263 -    Total Available RAM   : 24.00 GB
INFO:root:ArmoryUtils.py:1264 -    CPU ID string         : Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz

INFO:root:ArmoryUtils.py:1265 -    Number of CPU cores   : 8 cores
INFO:root:ArmoryUtils.py:1266 -    System is 64-bit      : True
INFO:root:ArmoryUtils.py:1267 -    Preferred Encoding    : UTF-8
INFO:root:ArmoryUtils.py:1268 -    Machine Arch          : x86_64
INFO:root:ArmoryUtils.py:1269 -    Available HDD (ARM)   : 173 GB
INFO:root:ArmoryUtils.py:1270 -    Available HDD (BTC)   : 173 GB
INFO:root:ArmoryUtils.py:1271 - 
INFO:root:ArmoryUtils.py:1272 - Network Name: Main Network
INFO:root:ArmoryUtils.py:1273 - Satoshi Port: 8333
INFO:root:ArmoryUtils.py:1274 - Do wlt check: True
INFO:root:ArmoryUtils.py:1275 - Named options/arguments to armoryengine.py:
INFO:root:ArmoryUtils.py:1277 -     thread_count    : -1
INFO:root:ArmoryUtils.py:1277 -     rescan          : False
INFO:root:ArmoryUtils.py:1277 -     ignoreAllZC     : False
INFO:root:ArmoryUtils.py:1277 -     rescanBalance   : False
INFO:root:ArmoryUtils.py:1277 -     disableModules  : False
INFO:root:ArmoryUtils.py:1277 -     port            : None
INFO:root:ArmoryUtils.py:1277 -     interport       : 8223
INFO:root:ArmoryUtils.py:1277 -     skipStatsReport : False
INFO:root:ArmoryUtils.py:1277 -     forceWalletCheck: False
INFO:root:ArmoryUtils.py:1277 -     regtest         : False
INFO:root:ArmoryUtils.py:1277 -     rebuild         : False
INFO:root:ArmoryUtils.py:1277 -     nettimeout      : 2
INFO:root:ArmoryUtils.py:1277 -     datadir         : DEFAULT
INFO:root:ArmoryUtils.py:1277 -     clearMempool    : False
INFO:root:ArmoryUtils.py:1277 -     offline         : False
INFO:root:ArmoryUtils.py:1277 -     coverageOutputDir: None
INFO:root:ArmoryUtils.py:1277 -     armoryDBDir     : DEFAULT
INFO:root:ArmoryUtils.py:1277 -     armorydb_port   : 9001
INFO:root:ArmoryUtils.py:1277 -     psn             : None
INFO:root:ArmoryUtils.py:1277 -     satoshiPort     : DEFAULT
INFO:root:ArmoryUtils.py:1277 -     useTorSettings  : False
INFO:root:ArmoryUtils.py:1277 -     netlog          : False
INFO:root:ArmoryUtils.py:1277 -     keypool         : 100
INFO:root:ArmoryUtils.py:1277 -     coverageInclude : None
INFO:root:ArmoryUtils.py:1277 -     forceOnline     : False
INFO:root:ArmoryUtils.py:1277 -     skipAnnounceCheck: False
INFO:root:ArmoryUtils.py:1277 -     redownload      : False
INFO:root:ArmoryUtils.py:1277 -     armorydb_ip     : 127.0.0.1
INFO:root:ArmoryUtils.py:1277 -     multisigFile    : DEFAULT
INFO:root:ArmoryUtils.py:1277 -     ram_usage       : -1
INFO:root:ArmoryUtils.py:1277 -     testAnnounceCode: False
INFO:root:ArmoryUtils.py:1277 -     mtdebug         : False
INFO:root:ArmoryUtils.py:1277 -     logDisable      : False
INFO:root:ArmoryUtils.py:1277 -     settingsPath    : /Users/Me/Library/Application Support/Armory/ArmorySettings.txt
INFO:root:ArmoryUtils.py:1277 -     db_type         : DB_FULL
INFO:root:ArmoryUtils.py:1277 -     doDebug         : False
INFO:root:ArmoryUtils.py:1277 -     enableDetSign   : True
INFO:root:ArmoryUtils.py:1277 -     disableConfPermis: False
INFO:root:ArmoryUtils.py:1277 -     testnet         : False
INFO:root:ArmoryUtils.py:1277 -     rpcport         : DEFAULT
INFO:root:ArmoryUtils.py:1277 -     satoshiHome     : DEFAULT
INFO:root:ArmoryUtils.py:1277 -     satoshiRpcport  : DEFAULT
INFO:root:ArmoryUtils.py:1277 -     logFile         : /Users/Me/Library/Application Support/Armory/ArmoryQt.py.log.txt
INFO:root:ArmoryUtils.py:1277 -     verbosity       : None
INFO:root:ArmoryUtils.py:1278 - Other arguments:
INFO:root:ArmoryUtils.py:1281 - ************************************************************
INFO:root:ArmoryUtils.py:1684 - C++ block utilities loaded successfully
INFO:root:BDM.py:367 - Using the asynchronous/multi-threaded BlockDataManager.
INFO:root:BDM.py:368 - Blockchain operations will happen in the background.  
INFO:root:BDM.py:369 - Devs: check TheBDM.getState() before asking for data.
INFO:root:BDM.py:370 - Registering addresses during rescans will queue them for 
INFO:root:BDM.py:371 - inclusion after the current scan is completed.
ERROR:root:Traceback (most recent call last):
  File "/Users/Me/Downloads/Armory 2.app/Contents/MacOS/py/usr/lib/armory/ArmoryQt.py", line 6891, in <module>
    import qt4reactor
  File "/Users/Me/Downloads/Armory 2.app/Contents/MacOS/py/usr/lib/armory/qt4reactor.py", line 44, in <module>
    from twisted.internet import posixbase
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/internet/posixbase.py", line 18, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/internet/udp.py", line 53, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/internet/address.py", line 14, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/python/filepath.py", line 16, in <module>
ImportError: cannot import name sha1

(ERROR) Traceback (most recent call last):
  File "/Users/Me/Downloads/Armory 2.app/Contents/MacOS/py/usr/lib/armory/ArmoryQt.py", line 6891, in <module>
    import qt4reactor
  File "/Users/Me/Downloads/Armory 2.app/Contents/MacOS/py/usr/lib/armory/qt4reactor.py", line 44, in <module>
    from twisted.internet import posixbase
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/internet/posixbase.py", line 18, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/internet/udp.py", line 53, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/internet/address.py", line 14, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/python/filepath.py", line 16, in <module>
ImportError: cannot import name sha1

Traceback (most recent call last):
  File "/Users/Me/Downloads/Armory 2.app/Contents/MacOS/py/usr/lib/armory/ArmoryQt.py", line 6891, in <module>
    import qt4reactor
  File "/Users/Me/Downloads/Armory 2.app/Contents/MacOS/py/usr/lib/armory/qt4reactor.py", line 44, in <module>
    from twisted.internet import posixbase
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/internet/posixbase.py", line 18, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/internet/udp.py", line 53, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/internet/address.py", line 14, in <module>
  File "/Users/Doug/Downloads/BitcoinArmory-master/osxbuild/workspace/Armory.app/Contents/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Twisted-16.3.0-py2.7-macosx-10.4-x86_64.egg/twisted/python/filepath.py", line 16, in <module>
ImportError: cannot import name sha1

`

Looks like dependencies issues are at work here.  Not sure who "/Users/Doug" is...

I suspect there is something in the development environment that was allowing it to run but is not present on a machine that does not have Brew installed.

**Comment by goatpig** [comment]
Doug is @droark, the OSX maintainer for Armory.

@utdrmac: Look into this thread for a possible fix: https://bitcointalk.org/index.php?topic=1732805.0

@TheCranston: you are missing the openSSL python package where twisted gets its crypto from. Ask @droark for more help if you can't get past the issue.

**Comment by TheCranston** [comment]
@goatpig Dang it..  I guess if I had read the README I'd see that I must install brew to make this work.  Brew totally fragged my machine and I had to clean reload because of it.  So I'm not very excited to have to load it again.  Is there any way that @droark could bundle the OpenSSL parts into the Armory.app package structure?  That would remove the dependency.  thanks for the consideration.

@utdrmac If you are willing to install brew (steps here: https://github.com/goatpig/BitcoinArmory/blob/dev/README_OSX.md ) and post your results that'd be awesome.

Thanks @goatpig!

**Comment by goatpig** [comment]
I'm not sure about the whole brew and twisted relationship, you'd have to ask @droark directly. I'm hoping to remove the need for twisted entirely in the 0.96 release phase, maybe that will fix your issue.

**Comment by utdrmac** [comment]
Sorry guys. I'm tagging out. Found a solution that fits my needs. Cheers.

**Comment by droark** [comment]
Hello. Sorry about all the confusion. The Mac build uses a strange hybrid of pre-bundled packages and brew-installed packages. This is due, in part, to some inadequacies in the versions of various packages included in OSX. It's basically the same build system that was present when I started contributing to Armory long ago. I'd like to update it. Time will tell. I believe there are some planned changes to the Armory build system. I'm reluctant to touch anything before that's settled.

Anyway, yes, #161 should handle one of the failures. The other is the OpenSSL thing. Honestly, I'm reluctant to include an OpenSSL package. If Armory relies on a specific library that turns out to have a major problem, the user could be screwed. Using brew works fine and gets around the incredibly ancient version shipped with OSX.

**Comment by oirman** [comment]
utdrmac...  I have the same problem and have tried everything I have come across in trying to solve it.  Can you remember how you got it working?

Thanks in advance...

**Comment by droark** [comment]
@oirman - What exactly is your problem? Which version of Armory are you trying to run? Have you read the [macOS README](https://github.com/goatpig/BitcoinArmory/blob/v0.96.3.991/README_macOS.md)?

**Comment by oirman** [comment]
Hi Doug,

Thanks for getting back to me on this.

I have tried to run everything from 93.1 upwards and am now attempting 
to use 0.96.3.991.  The furthest I've ever got is getting stuck about 
half way through verifying. 

After that happened many times, when tried to open it the icon would 
come up, flash and then disappear.  

Now nothing happens at all and I get the error message in my console: 

Jan 14 10:49:44 Bobbys-MacBook-Pro com.apple.xpc.launchd[1] 
(com.armory.armory.13276[4934]): Service exited with abnormal code: 126


This has all come off the back of my attempting to send coins with a 
too low fee in 93.1.  This was early last year and I've been trying to 
get it sorted ever since...

I have followed the instructions in the macOS README but to no avail.


Thanks again,

Tony
